sidecarhttp/http_api.go

```diff
@@ -0,0 +1,310 @@
+	response.Header().Set("Access-Control-Allow-Origin", "*")
+	response.Header().Set("Access-Control-Allow-Methods", "GET")
+	return
+}
+
+// watchHandler takes an optional GET parameter, "by_service"
+// By default, watchHandler returns `json.Marshal(state.ByService())` payloads
+// If the client passes "by_service=false", watchHandler returns `json.Marshal(state)` payloads
+func (s *SidecarApi) watchHandler(response http.ResponseWriter, req *http.Request, params map[string]string) {
+	defer req.Body.Close()
+
+	response.Header().Set("Content-Type", "application/json")
+
+	listener := NewHttpListener()
+
+	// Find out when the http connection closed so we can stop
+	notify := response.(http.CloseNotifier).CloseNotify()
+
+	// Let's subscribe to state change events
+	// AddListener and RemoveListener are thread safe
+	s.state.AddListener(listener)
+	defer s.state.RemoveListener(listener.Name())
+
+	byService := true
+	if req.URL.Query().Get("by_service") == "false" {
+		byService = false
+	}
+
+	var jsonBytes []byte
+	pushUpdate := func() error {
+		if byService {
+			s.state.RLock()
+			var err error
+			jsonBytes, err = json.Marshal(s.state.ByService())
+			s.state.RUnlock()
+
+			if err != nil {
+				return err
+			}
+		} else {
+			s.state.RLock()
+			jsonBytes = s.state.Encode()
+			s.state.RUnlock()
+		}
+
+		// In order to flush immediately, we have to cast to a Flusher.
+		// The normal HTTP library supports this but not all do, so we
+		// check just in case.
+		response.Write(jsonBytes)
+		if f, ok := response.(http.Flusher); ok {
+			f.Flush()
+		}
+
+		return nil
+	}
+
+	// Push the first update right away
+	err := pushUpdate()
+	if err != nil {
+		log.Errorf("Error marshaling state in watchHandler: %s", err.Error())
+		return
+	}
+
+	// Watch for further updates on the channel
+	for {
+		select {
+		case <-notify:
+			return
+
+		case <-listener.Chan():
+			err = pushUpdate()
+			if err != nil {
+				log.Errorf("Error marshaling state in watchHandler: %s", err.Error())
+				return
+			}
+		}
+	}
+}
+
+// oneServiceHandler takes the name of a single service and returns results for just
+// that service.
+func (s *SidecarApi) oneServiceHandler(response http.ResponseWriter, req *http.Request, params map[string]string) {
+	defer req.Body.Close()
+
+	response.Header().Set("Access-Control-Allow-Origin", "*")
+	response.Header().Set("Access-Control-Allow-Methods", "GET")
+	response.Header().Set("Content-Type", "application/json")
+
+	if params["extension"] != "json" {
+		sendJsonError(response, 404, "Not Found - Invalid content type extension")
+		return
+	}
+
+	name, ok := params["name"]
+	if !ok {
+		sendJsonError(response, 404, "Not Found - No service name provided")
+		return
+	}
+
+	if s.state == nil {
+		sendJsonError(response, 500, "Internal Server Error - Something went terribly wrong")
+		return
+	}
+
+	var instances []*service.Service
+	// Enter critical section
+	s.state.RLock()
+	defer s.state.RUnlock()
```

Comment:
Do we need to keep the lock for so long?

Reply:
We have to keep it until we've serialized the result. The other option is to copy the data structure, but that's slow, too. I think this is OK.